### DISCUSSION TITLE
Custom SDK Project: Restore in Visual Studio does not generate correct project.assets.json

### DISCUSSION THREAD

**Discussion by wasabii** [NuGet/Home#12439] [General]
I have a custom SDK project type named IKVM.NET.Sdk. It implements Java language support on .NET. It is largely a handle of additional targets added to Microsoft.NET.Sdk. As such, it ends up importing Microsoft.NET.Sdk Sdk.props and Sdk.targets.

I am having problems getting it to work in Visual Studio. It works fine in both dotnet build and msbuild.

The top-level problem is the .nuget.g.props and .nuget.g.targets files are never populated with user's PackageReference added projects that contain build and buildTransitive scripts. But, that's just a symptom of the deeper problem. On Restore, in Visual Studio, project.assets.json is written such that no PackageReference items are included. However, running dotnet restore includes them, as does msbuild /t:Restore.

So, you can run msbuild /t:Restore, everything is happy, then open Visual Studio, and bit Build (which does an implicit restore), and the file is overwritten incorrectly missing all the references. Fix the file. Build in Visual Studio, gone again.

It's very clearly something different in how Visual Studio does package restore vs how MSBuild does it. But this process seems opaque to me. I can't find any documentation on the differences, nor any debugging tool that might let me see the situation. There seems to be no way to generate binlogs of the MSBuild happening behind the scenesin VS during it's restore.... or whether it's even using MSBuild back there.

Any advice on this? Anything at all? Documentation? Insight into what differs in package restore in VS that might give me a hint as to the problem?

**Comment by zivkan** [comment]
> There seems to be no way to generate binlogs of the MSBuild happening behind the scenesin VS

The binlog viewer website has a link to the Visual Studio extension you can use to get binlogs out of VS: https://msbuildlog.com/#ProjectSystemTools

> Any advice on this?

NuGet doesn't run restore as a MSBuild target in VS. Historical thing, because for the first 5+ years that NuGet existed, PackageReference didn't exist, only packages.config, which has no msbuild integration. But it's still convenient not to run as an MSBuild target, because we can do up to date checks much faster this way compared to MSBuild, which helps with "change 1 line, run test" iteration speed.

For SDK style projects, VS normally uses https://github.com/dotnet/project-system, which is responsible for loading and parsing the MSBuild files, checking for all relevant data (what it calls a [Design Time Build](https://github.com/dotnet/project-system/blob/main/docs/design-time-builds.md)), then sending NuGet a notification via an API called `NominateProjectAsync` with all the PackageRefernces (and all other info that NuGet cares about). Restore works differently for non-SDK style project, but since your question is about an MSBuild SDK, only this project system is relevant.

Most likely cause is the Design Time Build (DTB) failing, which causes the project system to not know what PackageReferences exist. I no longer remember if the proejct system still sends a nomination when DTB fails. If NuGet doesn't receive a nomination for a project, I would have thought that it wouldn't create an assets file at all, but it's possible it creates a generic one with no packages. But without a nomination, NuGet won't know about project TFMs either.

Another thing worth investigating is turning on CPS debugging: https://github.com/dotnet/project-system/blob/9692d915612265e9635f328226f517e3bce45e01/docs/repo/debugging/logging.md?plain=1#L13
It outputs a lot of data, but in there is data that it sends to NuGet's NotifyProjectAsync API, so you can verify what it's telling NuGet about.

**Reply by wasabii** [reply]
DTBs succeed if I first run dotnet msbuild to generate the proper project.assets.json. Once I hit build in VS, which rewrites the project.assets.json file, DTBs no longer work (since nothing really works right after that).

The CPS debugging seems like a good place to start. I'll dig into that.

**Comment by TheXenocide** [comment]
Just as a possible consideration, does your Java project have a different file extension than, say, .csproj? Do you use a Visual Studio plugin for that project type? For example, WiX toolset v3 doesn't natively support SDK style projects. A while back we made a Wix SDK to take advantage of other features we've packaged in custom SDK format, but the Visual Studio experience is rather lackluster because it still uses the old-style project type handler for their wixproj extension in VS.

**Reply by wasabii** [reply]
It used to. was using .msbuildproj for a bit. Changed it to .csproj to see if it fixed it. It didn't.

No VS plugin currently.
